cadquery/occ_impl/solver.py

```diff
@@ -386,164 +551,173 @@ def fixed_rotation_axis_cost(m1: gp_Dir, t1: gp_Trsf, val: Tuple[int, float]):
     Fixed=fixed_cost,
     FixedPoint=fixed_point_cost,
     FixedAxis=fixed_axis_cost,
-    FixedRotationAxis=fixed_rotation_axis_cost,
+    FixedRotation=fixed_rotation_cost,
+)
+
+scaling: Dict[str, bool] = dict(
+    Point=True,
+    Axis=False,
+    PointInPlane=True,
+    PointOnLine=True,
+    Fixed=False,
+    FixedPoint=True,
+    FixedAxis=False,
+    FixedRotation=False,
 )
 
 # Actual solver class
 
 
 class ConstraintSolver(object):
 
-    entities: List[DOF6]
+    opti: ca.Opti
+    variables: List[Tuple[ca.MX, ca.MX]]  # variables
+    starting_points: List[Tuple[ca.MX, ca.MX]]  # variables
     constraints: List[Tuple[Tuple[int, ...], Constraint]]
     locked: List[int]
     ne: int
     nc: int
+    scale: float
 
     def __init__(
         self,
         entities: List[Location],
         constraints: List[Tuple[Tuple[int, ...], Constraint]],
         locked: List[int] = [],
+        scale: float = 1,
     ):
 
-        self.entities = [self._locToDOF6(loc) for loc in entities]
+        self.scale = scale
+        self.opti = opti = ca.Opti()
+        self.variables = [
+            (scale * opti.variable(NDOF_V), opti.variable(NDOF_Q))
+            if i not in locked
+            else (opti.parameter(NDOF_V), opti.parameter(NDOF_Q))
+            for i, _ in enumerate(entities)
+        ]
+        self.start_points = [
+            (opti.parameter(NDOF_V), opti.parameter(NDOF_Q)) for _ in entities
+        ]
+
+        # initialize, add the unit quaternion constraints and handle locked
+        for i, ((T, R), (T0, R0), loc) in enumerate(
+            zip(self.variables, self.start_points, entities)
+        ):
+            T0val, R0val = self._locToDOF6(loc)
+
+            opti.set_value(T0, T0val)
+            opti.set_value(R0, R0val)
+
+            if i in locked:
+                opti.set_value(T, (0, 0, 0))
+                opti.set_value(R, (0, 0, 0))
+            else:
+                opti.set_initial(T, (0.0, 0.0, 0.0))
+                opti.set_initial(R, (1e-2, 1e-2, 1e-2))
+
         self.constraints = constraints
 
         # additional book-keeping
         self.ne = len(entities)
         self.locked = locked
         self.nc = len(self.constraints)
 
-    @staticmethod
-    def _locToDOF6(loc: Location) -> DOF6:
+    def _locToDOF6(self, loc: Location) -> DOF6:
```

Comment:
What's the logic behind not having this be static anymore?

Reply:
Lost in translation, will fix